retrievalmarket/impl/client.go

```diff
@@ -79,18 +89,72 @@ func (c *client) Retrieve(ctx context.Context, pieceCID []byte, params retrieval
 			ID:       dealID,
 			Params:   params,
 		},
-		Status: retrievalmarket.DealStatusFailed,
-		Sender: miner,
+		TotalFunds:       totalFunds,
+		ClientWallet:     clientWallet,
+		MinerWallet:      minerWallet,
+		TotalReceived:    0,
+		CurrentInterval:  params.PaymentInterval,
+		BytesPaidFor:     0,
+		PaymentRequested: tokenamount.FromInt(0),
+		FundsSpent:       tokenamount.FromInt(0),
+		Status:           retrievalmarket.DealStatusNew,
+		Sender:           miner,
 	}
 
-	go func() {
-		evt := retrievalmarket.ClientEventError
-		c.notifySubscribers(evt, dealState)
-	}()
+	go c.handleDeal(ctx, dealState)
 
 	return dealID
 }
 
+func (c *client) failDeal(dealState *retrievalmarket.ClientDealState, err error) {
+	dealState.Message = err.Error()
+	dealState.Status = retrievalmarket.DealStatusFailed
+	c.notifySubscribers(retrievalmarket.ClientEventError, *dealState)
+}
+
+func (c *client) handleDeal(ctx context.Context, dealState retrievalmarket.ClientDealState) {
+
+	c.notifySubscribers(retrievalmarket.ClientEventOpen, dealState)
+
+	s, err := c.network.NewDealStream(dealState.Sender)
+	if err != nil {
+		c.failDeal(&dealState, err)
+		return
+	}
+	defer s.Close()
+
+	environment := clientDealEnvironment{c.node, &UnixFs0Verifier{Root: dealState.DealProposal.PayloadCID}, c.bs, s}
+
+	for {
+		var handler clientstates.ClientHandlerFunc
+
+		switch dealState.Status {
+		case retrievalmarket.DealStatusNew:
+			handler = clientstates.ProposeDeal
+		case retrievalmarket.DealStatusAccepted:
+			handler = clientstates.SetupPaymentChannel
+		case retrievalmarket.DealStatusPaymentChannelCreated, retrievalmarket.DealStatusOngoing, retrievalmarket.DealStatusUnsealing:
+			handler = clientstates.ProcessNextResponse
+		case retrievalmarket.DealStatusFundsNeeded, retrievalmarket.DealStatusFundsNeededLastPayment:
+			handler = clientstates.ProcessNextResponse
+		default:
+			c.failDeal(&dealState, errors.New("unexpected deal state"))
+			return
+		}
+		dealModifier := handler(ctx, environment, dealState)
+		dealModifier(&dealState)
+		if retrievalmarket.IsTerminalStatus(dealState.Status) {
+			break
+		}
+		c.notifySubscribers(retrievalmarket.ClientEventProgress, dealState)
+	}
+	if retrievalmarket.IsTerminalSuccess(dealState.Status) {
```

Comment:
So the status at this line must be terminal -- cause the break in the above loop requires that.

And that notification for status inside the loop is below the break so that the "in progress" notification only goes out if the status is not terminal after applying dealModifier.

my notion of how things will proceed in the loop based on the code above is:
- at start of loop, status is non terminal
- apply logic based on starting status
- apply modifier for new status
- check if new status is terminal, if so exit loop
- notify in progress status update and start loop again, but only if status is non-terminal

and the code at the bottom is:
with terminal status that exited loop--
- if success notify complete
- if terminal error, notify error

Does that track with your understanding of this code?

Reply:
Ok that makes sense.